File: unitypomelo-server/game-server/app/domain/action/actionManager.js
var Queue = require('pomelo-collection').queue;
var logger = require('pomelo-logger').getLogger(__filename);

var ActionManager =function(opts){
	opts = opts||{};
	this.limit = opts.limit || 10000;
	//行为映射字典: [type][id]=action
	this.actionMap = {};
	//行为队列，所有的行为以FIFO方式执行
	this.actionQueue = new Queue(this.limit);
	
};

var pro = ActionManager.prototype;
pro.addAction=function(action){
	if(action.singleton) this.abortAction(action.type,action.id);
	
	this.actionMap[action.type] = this.actionMap[action.type]||{};
	this.actionMap[action.type][action.id] =action;
	return this.actionQueue.push(action);
};

pro.abortAction = function(type,id){
	if(!this.actionMap[type] || !this.actionMap[type][id]) return;
	this.actionMap[type][id].aborted = true;
	delete this.actionMap[type][id];
};

pro.abortAllAction = function(id){
	for(var type in this.actionMap){
		if(!!this.actionMap[type][id]) {
			this.actionMap[type][id].aborted = true;
		}
	}
};

pro.update = function(){
	var length = this.actionQueue.length;
	for(var i =0;i<length;i++){
		var action = this.actionQueue.pop();
		if(action.aborted) continue;
		
		action.update();
		if(!action.finished)
			this.actionQueue.push(action);
		else
			delete this.actionMap[action.type][action.id];
	}
};

module.exports=ActionManager;

File: UnityPomelo-client/Assets/Scripts/Action/ActionManager.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System;

public class ActionManager
{
    private int limit = 1000;
    private Dictionary<string, Dictionary<int, BaseAction>> actionMap;
    private Queue<BaseAction> actionQueue;

    public ActionManager(int limit =0)
    {
        this.limit = Mathf.Max(limit, this.limit);
        this.actionMap = new Dictionary<string, Dictionary<int,BaseAction>>();
        this.actionQueue = new Queue<BaseAction>(this.limit);
    }

    public void addAction(BaseAction action)
    {
        if (action.singleton || 
            (this.actionMap.ContainsKey(action.type) && this.actionMap[action.type].ContainsKey(action.id)))
            this.abortAction(action.type, action.id);
  
        if (!this.actionMap.ContainsKey(action.type))
        {
            var dict = new Dictionary<int, BaseAction>();
            dict.Add(action.id, action);
            this.actionMap.Add(action.type, dict);
        }else
        {
            if (!this.actionMap[action.type].ContainsKey(action.id))
            {
                this.actionMap[action.type].Add(action.id, action);
            }
        }

        this.actionQueue.Enqueue(action);

    }

    public void abortAction(string type,int id)
    {
        if (!this.actionMap.ContainsKey(type) || !this.actionMap[type].ContainsKey(id)) return;

        this.actionMap[type][id].aborted = true;
        this.actionMap[type].Remove(id);
    }

    public void abortAllAction(int id)
    {
        foreach(var p in this.actionMap)
        {
            if (p.Value.ContainsKey(id)) p.Value[id].aborted = true;
        }
    }

    public void update()
    {
        var length = this.actionQueue.Count;
        if (length == 0) return;
        var action = this.actionQueue.Dequeue();
        if (action.aborted) return;
        action.update();

        if (!action.finished) this.actionQueue.Enqueue(action);
        else this.actionMap[action.type].Remove(action.id);
        
    }
}
